pandas/core/internals.py

```diff
@@ -2513,7 +2513,7 @@ def putmask(self, **kwargs):
         return self.apply('putmask', **kwargs)
 
     def diff(self, **kwargs):
-        return self.apply('diff', **kwargs)
+        return self.consolidate().apply('diff', **kwargs)
```

Comment:
instead of this

add to ``.apply`` an argument ``consolidate=True``, and:
```
if consolidate:
    self._consolidate_inplace()
```
This should fix this issue and not break anything else.